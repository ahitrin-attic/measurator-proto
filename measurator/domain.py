import datetime


class IO:
    """Custom wrapper for all input/output interactions"""

    def write(self, text: str, *args) -> None:
        raise NotImplementedError

    def read(self) -> str:
        raise NotImplementedError

    def write_file(self, data: iter) -> None:
        raise NotImplementedError

    def read_file(self) -> iter:
        raise NotImplementedError

    def now(self) -> datetime.datetime:
        raise NotImplementedError
