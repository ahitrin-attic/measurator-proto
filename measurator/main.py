import argparse
import csv
import datetime
import time


def migrate():
    pass


def run_main():
    path = _file_path()
    not_yet, succeeds, fails = _read_file(path)
    # evaluate measurements
    now = datetime.datetime.now()
    delayed = list()
    for status, timestamp, text in not_yet:
        evaluate_time = datetime.datetime(*(time.strptime(timestamp, "%Y-%m-%d %H:%M")[:6]))
        if evaluate_time < now:
            print("Time to evaluate: {}\n Is it true? (Yes/No/Delay)".format(text))
            user_input = input().capitalize()
            if user_input.startswith("Y"):
                status = "S"
                succeeds.append((status, timestamp, text))
            elif user_input.startswith("D"):
                print("When to evaluate (YYYY-mm-dd HH:MM):")
                eval_time = input()
                delayed.append(("N", eval_time, text))
            else:
                status = "F"
                fails.append((status, timestamp, text))
    not_yet = delayed
    # print total statistics
    total_done = len(fails) + len(succeeds)
    if total_done > 0:
        percentage = "%d%%" % (float(100 * len(succeeds)) / float(total_done))
    else:
        percentage = "N/A"
    print("Succesful predictions:", percentage, ", not done yet:", len(not_yet))
    # add another prediction when needed
    print("Add another prediction? Y/N")
    user_input = input()
    if user_input.capitalize().startswith("Y"):
        print("Prediction:")
        prediction = input()
        print("When to evaluate (YYYY-mm-dd HH:MM):")
        eval_time = input()
        not_yet.append(("N", eval_time, prediction))
    # overwrite predictions file
    all_rows = list()
    all_rows.extend(fails)
    all_rows.extend(succeeds)
    all_rows.extend(not_yet)
    with open(path, "wt") as f:
        writer = csv.writer(f)
        for row in all_rows:
            writer.writerow(row)


def _read_file(path):
    fails = list()
    succeeds = list()
    not_yet = list()
    try:
        with open(path) as f:
            reader = csv.reader(f)
            for row in reader:
                status = row[0]
                if status == "F":
                    fails.append(row)
                elif status == "S":
                    succeeds.append(row)
                else:
                    not_yet.append(row)
    except FileNotFoundError:
        pass
    return not_yet, succeeds, fails


def _file_path():
    parser = argparse.ArgumentParser()
    parser.add_argument("path")
    args = parser.parse_args()
    return args.path
