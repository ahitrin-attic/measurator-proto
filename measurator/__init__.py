from measurator.main import run_main, migrate
