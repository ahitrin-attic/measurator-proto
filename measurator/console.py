import datetime

from measurator.domain import IO


class ConsoleIO(IO):
    def __init__(self, path) -> None:
        super().__init__()
        self._path = path

    def write(self, text: str, *args):
        print(text, *args)

    def read(self) -> str:
        return input()

    def write_file(self, data: iter) -> None:
        with open(self._path, "w") as f:
            for line in data:
                f.write(line)

    def read_file(self) -> iter:
        try:
            with open(self._path) as f:
                return f.readlines()
        except FileNotFoundError:
            return []

    def now(self) -> datetime.datetime:
        return datetime.datetime.now()
