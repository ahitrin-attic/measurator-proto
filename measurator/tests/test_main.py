import datetime
from typing import List

from approvaltests import verify
from approvaltests.reporters import GenericDiffReporterFactory

from measurator import run_main_
from measurator.domain import IO


class TestableIO(IO):
    def __init__(self, inputs: List[str], timestamps: List[datetime.datetime], file_content: iter):
        self.file_content = file_content
        self.inputs = inputs
        self.timestamps = timestamps
        self.log: List[str] = []

    def write(self, text: str, *args) -> None:
        self.log.append(f" {text}{''.join(str(x) for x in args)}")

    def read(self) -> str:
        output = self.inputs.pop()
        self.log.append(f"> {output}")
        return output

    def write_file(self, data: iter) -> None:
        self.log.append(f"! {'|'.join(data)}")

    def read_file(self) -> iter:
        self.log.append("* READ FILE")
        return self.file_content

    def now(self) -> datetime.datetime:
        timestamp = self.timestamps.pop()
        self.log.append(f"@ {timestamp}")
        return timestamp


def test_approve():
    inputs = ['N']
    timestamps = []
    file_content = []
    io = TestableIO(inputs, timestamps, file_content)
    run_main_(io)
    reporter = GenericDiffReporterFactory().get_first_working()
    verify("\n".join(io.log), reporter)
