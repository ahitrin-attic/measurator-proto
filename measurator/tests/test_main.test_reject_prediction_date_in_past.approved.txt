* READ FILE
Current time is 2020-04-01 12:00
 Successful predictions:N/A, not done yet:0
 Add another prediction? Yes/*No*/List
> Y
 Prediction:
> prediction
 When to evaluate (YYYY-mm-dd HH:MM):
> 2020-04-01 11:59
 This date is in past, prediction is not saved!
File is written:
```

```